From 704a6cd01e65ed82536cc33960b6401efd7a7514 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 02:16:14 -0400
Subject: [pci] Add support for PCIe-to-PCI bridge DMA alias quirks

Message-id: <20140905021614.32574.10484.stgit@gir.stowe>
Patchwork-id: 91365
O-Subject: [RHEL7.1 PATCH 06/17] PCI: Add support for PCIe-to-PCI bridge DMA alias quirks
Bugzilla: 1136537
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1136537
Upstream Status: c8fe16e3f96a9bb95a10cedb19d2be2d2d580940

commit c8fe16e3f96a9bb95a10cedb19d2be2d2d580940
Author: Alex Williamson <alex.williamson@redhat.com>
Date:   Wed May 28 14:57:02 2014 -0600

    PCI: Add support for PCIe-to-PCI bridge DMA alias quirks

    Several PCIe-to-PCI bridges fail to provide a PCIe capability, causing us
    to handle them as conventional PCI devices when they really use the
    requester ID of the secondary bus.  We need to differentiate these from
    PCIe-to-PCI bridges that actually use the conventional PCI ID when a PCIe
    capability is not present, such as those found on the root complex of may
    Intel chipsets.  Add a dev_flag bit to identify devices to be handled as
    standard PCIe-to-PCI bridges.

    Signed-off-by: Alex Williamson <alex.williamson@redhat.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pci/search.c b/drivers/pci/search.c
index d02e2d1..3966fb1 100644
--- a/drivers/pci/search.c
+++ b/drivers/pci/search.c
@@ -89,8 +89,14 @@ int pci_for_each_dma_alias(struct pci_dev *pdev,
     continue;
    }
   } else {
-   ret = fn(tmp, PCI_DEVID(tmp->bus->number, tmp->devfn),
-     data);
+   if (tmp->dev_flags & PCI_DEV_FLAG_PCIE_BRIDGE_ALIAS)
+    ret = fn(tmp,
+      PCI_DEVID(tmp->subordinate->number,
+         PCI_DEVFN(0, 0)), data);
+   else
+    ret = fn(tmp,
+      PCI_DEVID(tmp->bus->number,
+         tmp->devfn), data);
    if (ret)
     return ret;
   }
diff --git a/include/linux/pci.h b/include/linux/pci.h
index 7893330..020ac2d 100644
--- a/include/linux/pci.h
+++ b/include/linux/pci.h
@@ -173,6 +173,8 @@ enum pci_dev_flags {
  PCI_DEV_FLAGS_ACS_ENABLED_QUIRK = (__force pci_dev_flags_t) (1 << 3),
  /* Flag to indicate the device uses dma_alias_devfn */
  PCI_DEV_FLAGS_DMA_ALIAS_DEVFN = (__force pci_dev_flags_t) (1 << 4),
+ /* Use a PCIe-to-PCI bridge alias even if !pci_is_pcie */
+ PCI_DEV_FLAG_PCIE_BRIDGE_ALIAS = (__force pci_dev_flags_t) (1 << 5),
 };
 
 enum pci_irq_reroute_variant {
-- 
1.7.1